File: app/services/Legacy/OllamaClient.php
<?php

namespace App\Services\Legacy;
/**
 * Ollama Client for Local AI Integration
 * Provides a simple interface to interact with Ollama's local API
 */

class OllamaClient {
    private $baseUrl;
    private $model;

    public function __construct($model = 'llama3', $baseUrl = 'http://localhost:11434') {
        $this->baseUrl = $baseUrl;
        $this->model = $model;
    }

    /**
     * Generate text using Ollama
     * @param string $prompt The input prompt
     * @param array $options Additional options (temperature, etc.)
     * @return string|null Generated text or null on error
     */
    public function generate($prompt, $options = []) {
        $data = [
            'model' => $this->model,
            'prompt' => $prompt,
            'stream' => false
        ];

        if (!empty($options)) {
            $data = array_merge($data, $options);
        }

        $response = $this->makeRequest('/api/generate', $data);

        if ($response && isset($response['response'])) {
            return $response['response'];
        }

        return null;
    }

    /**
     * Chat with the model
     * @param array $messages Array of message objects
     * @param array $options Additional options
     * @return string|null Response or null on error
     */
    public function chat($messages, $options = []) {
        $data = [
            'model' => $this->model,
            'messages' => $messages,
            'stream' => false
        ];

        if (!empty($options)) {
            $data = array_merge($data, $options);
        }

        $response = $this->makeRequest('/api/chat', $data);

        if ($response && isset($response['message']['content'])) {
            return $response['message']['content'];
        }

        return null;
    }

    /**
     * Check if Ollama is running
     * @return bool
     */
    public function isAvailable() {
        $response = $this->makeRequest('/api/tags', []);
        return $response !== null;
    }

    /**
     * List available models
     * @return array|null
     */
    public function listModels() {
        $response = $this->makeRequest('/api/tags', []);
        return $response ? $response['models'] : null;
    }

    /**
     * Make HTTP request to Ollama API
     * @param string $endpoint API endpoint
     * @param array $data Request data
     * @return array|null Decoded JSON response or null on error
     */
    private function makeRequest($endpoint, $data) {
        $url = $this->baseUrl . $endpoint;

        $ch = curl_init($url);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_POST, true);
        curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($data));
        curl_setopt($ch, CURLOPT_HTTPHEADER, [
            'Content-Type: application/json'
        ]);
        curl_setopt($ch, CURLOPT_TIMEOUT, 30); // Timeout for local requests

        $response = curl_exec($ch);
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);

        if ($httpCode === 200 && $response) {
            return json_decode($response, true);
        }

        return null;
    }
}

/**
 * Helper function to get Ollama client instance
 * @param string $model Model name (default: llama3)
 * @return OllamaClient
 */
function getOllamaClient($model = 'llama3') {
    return new OllamaClient($model);
}
?>

File: agentic_dev_system/py_agentic/ollama_client.py
"""
Ollama Client - Lightweight HTTP client for Ollama API.

Uses only Python standard library (urllib) - no external dependencies.
Communicates with local Ollama server for AI reasoning tasks.

Usage:
    client = OllamaClient({'host': 'localhost', 'port': 11434, 'model': 'qwen2.5:7b'})
    if client.is_available():
        response = client.generate("Fix this PHP code: ...")
"""

import json
import urllib.request
import urllib.error
from typing import Dict, Any, Optional


class OllamaClient:
    """Lightweight Ollama API client using stdlib only."""

    def __init__(self, config: dict):
        self.host = config.get('host', 'localhost')
        self.port = config.get('port', 11434)
        self.model = config.get('model', 'qwen2.5:7b')
        self.base_url = f"http://{self.host}:{self.port}"
        self.timeout = config.get('timeout', 30)

    def is_available(self) -> bool:
        """Check if Ollama server is reachable."""
        try:
            req = urllib.request.Request(f"{self.base_url}/api/tags")
            with urllib.request.urlopen(req, timeout=3) as resp:
                return resp.status == 200
        except Exception:
            return False

    def generate(self, prompt: str, system: str = '', max_tokens: int = 4096) -> str:
        """
        Generate a response from Ollama.

        Args:
            prompt: The user prompt
            system: System prompt for context
            max_tokens: Maximum tokens to generate

        Returns:
            Generated text response
        """
        payload = {
            'model': self.model,
            'prompt': prompt,
            'stream': False,
            'options': {
                'temperature': 0.7,
                'num_predict': max_tokens
            }
        }

        if system:
            payload['system'] = system

        try:
            data = json.dumps(payload).encode('utf-8')
            req = urllib.request.Request(
                f"{self.base_url}/api/generate",
                data=data,
                headers={'Content-Type': 'application/json'}
            )
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                result = json.loads(resp.read().decode('utf-8'))
                return result.get('response', '')
        except Exception as e:
            return f"[Ollama error: {e}]"

    def chat(self, messages: list) -> str:
        """
        Chat completion with message history.

        Args:
            messages: List of {"role": "user"/"assistant"/"system", "content": "..."}

        Returns:
            Assistant response text
        """
        payload = {
            'model': self.model,
            'messages': messages,
            'stream': False,
            'options': {
                'temperature': 0.7
            }
        }

        try:
            data = json.dumps(payload).encode('utf-8')
            req = urllib.request.Request(
                f"{self.base_url}/api/chat",
                data=data,
                headers={'Content-Type': 'application/json'}
            )
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                result = json.loads(resp.read().decode('utf-8'))
                msg = result.get('message', {})
                return msg.get('content', '')
        except Exception as e:
            return f"[Ollama error: {e}]"
